### PR TITLE
[DataGrid] Support `minHeight` and `maxHeight` on flex parent container

### PR DESCRIPTION
**Pull request by cherniavskii**
Closes https://github.com/mui/mui-x/issues/3524
Closes https://github.com/mui/mui-x/issues/7588
Closes https://github.com/mui/mui-x/issues/9178
Closes https://github.com/mui/mui-x/pull/9274

Preview:
- https://deploy-preview-14614--material-ui-x.netlify.app/x/react-data-grid/layout/#flex-parent-container
- https://deploy-preview-14614--material-ui-x.netlify.app/x/react-data-grid/layout/#minimum-and-maximum-height

### TODO
- [x] Add regression tests
- [x] Deprecate `autoHeight` prop?

### CODE REVIEW

File: docs/data/data-grid/layout/MinMaxHeightGrid.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        <ContainerMeasurements containerRef={containerRef} />
+      </div>
+    </Box>
+  );
+}
+
+function ContainerMeasurements({ containerRef }) {
+  const [containerHeight, setContainerHeight] = React.useState(0);
+
+  useResizeObserver(containerRef, (entries) => {
```
Comment (cherniavskii) [inline docs/data/data-grid/layout/MinMaxHeightGrid.js:57]: I added this component to visualize the DataGrid height (no need to check it in the dev tools):



It works fine locally, but the resize observer callback is not called in the deploy preview for some reason 🫠
I'll look into it later, but the PR in its current form is ready for review.

File: docs/data/data-grid/layout/MinMaxHeightGrid.js
Comment (romgrk) [inline docs/data/data-grid/layout/MinMaxHeightGrid.js:1]: Small issue with the scrollbar on chrome/linux (which usually behaves like chrome/windows):

[Screencast from 2024-09-18 17-48-40.webm](https://github.com/user-attachments/assets/99e9078b-6278-471f-94d0-618e15baa97a)

File: packages/x-data-grid/src/components/virtualization/GridBottomContainer.tsx [hunk 2]
```diff
@@ -20,10 +23,30 @@ const Element = styled('div')({
...
+      minimumSize.height -
+      // Subtract scroll height twice to account for GridVirtualScrollerFiller and horizontal scrollbar
+      2 * scrollHeight,
+    0,
+  );
+
+  let style = props.style;
+  if (offset !== 0) {
+    style = { ...style, transform: `translateY(${offset}px)` };
+  }
```
Comment (romgrk) [inline packages/x-data-grid/src/components/virtualization/GridBottomContainer.tsx:43]: We don't use the HTML div props from the top/bottom container, we could remove the props parameter and avoid this block of code.

File: docs/data/data-grid/layout/MinMaxHeightGrid.js
Comment (cherniavskii) [reply docs/data/data-grid/layout/MinMaxHeightGrid.js:1]: Hmm, I can't reproduce it in Chromium on Ubuntu:


https://github.com/user-attachments/assets/31fae484-0588-4582-aec4-985348ccb1a2

Comment (romgrk) [reply docs/data/data-grid/layout/MinMaxHeightGrid.js:1]: It only happens when the zoom is not 100%.

Comment (cherniavskii) [reply docs/data/data-grid/layout/MinMaxHeightGrid.js:1]: Ah, with non-100% zoom, I can reproduce this on MacOS as well 👍🏻
Will look into it!

File: packages/x-data-grid/src/components/virtualization/GridVirtualScrollbar.tsx [hunk 3]
```diff
@@ -69,9 +66,7 @@ const ScrollbarHorizontal = styled(Scrollbar)({
   bottom: '0px',
 });
 
-const Content = styled('div')({
-  display: 'inline-block',
```
Comment (cherniavskii) [inline packages/x-data-grid/src/components/virtualization/GridVirtualScrollbar.tsx:73]: Not sure why we needed it, but removing it also fixes the issue I just noticed:

Before: https://mui.com/x/react-data-grid/layout/#predefined-dimensions

https://github.com/user-attachments/assets/9c770446-d122-4e45-a0a8-f9c763eef338

After: https://deploy-preview-14614--material-ui-x.netlify.app/x/react-data-grid/layout/#predefined-dimensions

https://github.com/user-attachments/assets/6aa171f6-f81d-4420-a140-f515bdfa6d54

File: docs/data/data-grid/layout/layout.md [hunk 4]
```diff
@@ -8,6 +8,20 @@ Instead, it takes up the space given by its parent.
 The Data Grid will raise an error in the console if its container has no intrinsic dimensions.
 :::
 
+## Flex parent container
+
+The Data Grid can be placed inside a flex container with `flex-direction: column`.
+Without setting the minimum and maximum height, the Data Grid will take as much space as it needs to display all rows.
+This is effectively equivalent to setting `autoHeight={true}`, but with the added benefit of being able to set the minimum and maximum height of the parent container.
```
Comment (cherniavskii) [inline docs/data/data-grid/layout/layout.md:15]: Given the above, I feel like we should deprecate the `autoHeight` prop now.
What do you think?

File: packages/x-data-grid/src/components/virtualization/GridBottomContainer.tsx (hunk 2 above)
Comment (cherniavskii) [reply packages/x-data-grid/src/components/virtualization/GridBottomContainer.tsx]: Done for both top and bottom containers 👍🏻

File: docs/data/data-grid/layout/MinMaxHeightGrid.js
Comment (cherniavskii) [reply docs/data/data-grid/layout/MinMaxHeightGrid.js:1]: Fixed, and the fix also solves another issue that I didn't notice before: https://github.com/mui/mui-x/pull/14614#discussion_r1766882578

File: docs/data/data-grid/layout/layout.md (hunk 4 above)
Comment (romgrk) [reply docs/data/data-grid/layout/layout.md:15]: Sounds good to me.

**Review by romgrk — APPROVED** [review]

File: packages/x-data-grid/src/components/virtualization/GridVirtualScrollbar.tsx [hunk 5]
```diff
@@ -69,9 +66,7 @@ const ScrollbarHorizontal = styled(Scrollbar)({
...
-});
+const Content = styled('div')(null);
```
Comment (KenanYusuf) [inline packages/x-data-grid/src/components/virtualization/GridVirtualScrollbar.tsx:69]: Could we remove this if there are no styles to add?

File: docs/data/data-grid/layout/layout.md (hunk 4 above)
Comment (KenanYusuf) [inline docs/data/data-grid/layout/layout.md:15]: This line could be placed in a callout below the demo with a link to the **Auto height** section of the page.

File: docs/data/data-grid/layout/layout.md [hunk 6]
```diff
@@ -8,6 +8,20 @@ Instead, it takes up the space given by its parent.
...
+
+{{"demo": "FlexGrid.js", "bg": "inline"}}
+
+### Minimum and maximum height
+
+In the demo below, the Data Grid is placed inside a flex container with a minimum height of 200px and a maximum height of 400px and adapts its height when the number of rows changes.
```
Comment (KenanYusuf) [inline docs/data/data-grid/layout/layout.md:21]:
```suggestion
In the demo below, the Data Grid is placed inside a flex container with a minimum height of `200px` and a maximum height of `400px` and adapts its height when the number of rows changes.
```

Should the values be formatted?

File: docs/data/data-grid/layout/layout.md (hunk 4 above)
Comment (cherniavskii) [reply docs/data/data-grid/layout/layout.md]: Done in https://deploy-preview-14614--material-ui-x.netlify.app/x/react-data-grid/layout/#flex-parent-container

Comment (cherniavskii) [reply docs/data/data-grid/layout/layout.md]: Done 👍🏻

File: docs/data/data-grid/layout/MinMaxHeightGrid.js (hunk 1 above)
Comment (cherniavskii) [reply docs/data/data-grid/layout/MinMaxHeightGrid.js:57]: Fixed by [`14cbefa` (#14614)](https://github.com/mui/mui-x/pull/14614/commits/14cbefa7ed55269cf30ee8a06830ca37a250c145)

**Review by KenanYusuf — APPROVED** [review]

**Review by owenchak — APPROVED** [review]

File: docs/data/data-grid/layout/MinMaxHeightGrid.js
Comment (michelengelen) [reply docs/data/data-grid/layout/MinMaxHeightGrid.js:1]: Could this be related to #14814 @romgrk ?

Comment (cherniavskii) [reply docs/data/data-grid/layout/MinMaxHeightGrid.js:1]: I don't think so, I can reproduce the issue with both 7.19.0 and 7.18.0 (before this PR was merged)

**Pull request merged**